spec/java/util/stream/DoubleStream.main.lsl

```diff
@@ -0,0 +1,1266 @@
+        {
+            var newArray: array<double> = action ARRAY_NEW("double", 0);
+            result = new DoubleStreamAutomaton(state = Initialized,
+                storage = newArray,
+                length = 0,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+        else
+        {
+            // what will be if will be overflow ?
+            val offset: int = n as int;
+            val newLength: int = this.length - offset;
+            val skipStorage: array<double> = action ARRAY_NEW("double", newLength);
+
+            var i: int = 0;
+            var skipIndex: int = 0;
+            action LOOP_FOR(
+                i, offset, this.length, +1,
+                _skip_loop(i, skipIndex, skipStorage)
+            );
+
+            result = new DoubleStreamAutomaton(state = Initialized,
+                storage = skipStorage,
+                length = newLength,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+
+        this.linkedOrConsumed = true;
+    }
+
+
+    @Phantom proc _skip_loop (i: int, skipIndex: int, skipStorage: array<double>): void
+    {
+        skipStorage[skipIndex] = this.storage[i];
+        skipIndex += 1;
+    }
+
+
+    fun *.forEach (@target self: DoubleStream, _action: DoubleConsumer): void
+    {
+        if (this.linkedOrConsumed)
+            _throwISE();
+
+        _actionApply(_action);
+        this.linkedOrConsumed = true;
+    }
+
+
+    fun *.forEachOrdered (@target self: DoubleStream, _action: DoubleConsumer): void
+    {
+        if (this.linkedOrConsumed)
+            _throwISE();
+
+        _actionApply(_action);
+        this.linkedOrConsumed = true;
+    }
+
+
+    fun *.toArray (@target self: DoubleStream): array<double>
+    {
+        if (this.linkedOrConsumed)
+            _throwISE();
+
+        result = this.storage;
+        this.linkedOrConsumed = true;
+    }
+
+
+    fun *.reduce (@target self: DoubleStream, identity: double, accumulator: DoubleBinaryOperator): double
+    {
+        if (this.linkedOrConsumed)
+            _throwISE();
+
+        if (accumulator == null)
+            _throwNPE();
+
+        result = identity;
+
+        if (this.length != 0)
+        {
+            action ASSUME(this.length > 0);
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 0, this.length, +1,
+                _accumulate_loop(i, accumulator, result)
+            );
+        }
+
+        this.linkedOrConsumed = true;
+    }
+
+
+    @Phantom proc _accumulate_loop (i: int, accumulator: DoubleBinaryOperator, result: double): void
+    {
+        result = action CALL(accumulator, [result, this.storage[i]]);
+    }
+
+
+    fun *.reduce (@target self: DoubleStream, accumulator: DoubleBinaryOperator): OptionalDouble
+    {
+        if (this.linkedOrConsumed)
+            _throwISE();
+
+        if (accumulator == null)
+            _throwNPE();
+
+        var value: double = 0;
```

Comment:
Why we need variable here but not in one of the branches of the `if` statement down below?
*And I know, this is a micro-optimization but still.*

Reply:
Fixed;